packages/zoltan2/core/src/algorithms/color/Zoltan2_AlgHybrid2GL.hpp

```diff
@@ -0,0 +1,1532 @@
+      //rand has an entry for each vertex, so its size corresponds to what is needed.
+      Kokkos::View<lno_t*, device_type> verts_to_recolor_view("verts to recolor", rand.size());
+      typename Kokkos::View<lno_t*, device_type>::HostMirror verts_to_recolor_host = create_mirror(verts_to_recolor_view);
+      Kokkos::View<lno_t*, device_type, Kokkos::MemoryTraits<Kokkos::Atomic>> verts_to_recolor_atomic = verts_to_recolor_view;
+      
+      //This view keeps track of the size of the list of vertices to recolor.
+      Kokkos::View<int*, device_type> verts_to_recolor_size("verts to recolor size",1);
+      Kokkos::View<int*, device_type, Kokkos::MemoryTraits<Kokkos::Atomic>> verts_to_recolor_size_atomic = verts_to_recolor_size;
+      typename Kokkos::View<int*, device_type>::HostMirror verts_to_recolor_size_host = create_mirror(verts_to_recolor_size);
+      
+      //initialize the host view
+      verts_to_recolor_size_host(0) = 0;
+      //copy to device
+      Kokkos::deep_copy(verts_to_recolor_size, verts_to_recolor_size_host);
+
+      //verts to send can only include locally owned vertices,
+      //so we can safely allocate the view to size n_local.
+      //
+      //This view is a list of local vertices that are going to be
+      //recolored, and need to be sent to their remote copies.
+      Kokkos::View<lno_t*, device_type> verts_to_send_view("verts to send", n_local);
+      Kokkos::View<lno_t*, device_type, Kokkos::MemoryTraits<Kokkos::Atomic>> verts_to_send_atomic = verts_to_send_view;
+      typename Kokkos::View<lno_t*, device_type>::HostMirror verts_to_send_host = create_mirror(verts_to_send_view);
+      
+      //this view keeps track of the size of verts_to_send.
+      Kokkos::View<size_t*, device_type> verts_to_send_size("verts to send size",1);
+      Kokkos::View<size_t*, device_type, Kokkos::MemoryTraits<Kokkos::Atomic>> verts_to_send_size_atomic = verts_to_send_size;
+      typename Kokkos::View<size_t*, device_type>::HostMirror verts_to_send_size_host = create_mirror(verts_to_send_size);
+
+      verts_to_send_size_host(0) = 0;
+      Kokkos::deep_copy(verts_to_send_size, verts_to_send_size_host);
+
+      if(verbose) std::cout<<comm->getRank()<<": Constructing the boundary\n";
+      
+      //this function allocates and initializes the boundary_verts_dev view,
+      //sets the boundary_size accordingly, and adds vertices to the
+      //verts_to_send_atomic view, updating the size view as well.
+      constructBoundary(n_local, dist_offsets_dev, dist_adjs_dev, dist_offsets_host, dist_adjs_host, boundary_verts_dev, 
+	                boundary_size, verts_to_send_atomic, verts_to_send_size_atomic);
+      
+      //this boolean chooses which KokkosKernels algorithm to use.
+      //It was experimentally chosen for distance-1 coloring.
+      //Should be disregarded for distance-2 colorings.
+      bool use_vbbit = (global_max_degree < 6000);
+      //Done initializing, start coloring!
+      comm->barrier();
```

Comment:
Is barrier needed only for timing?

Reply:
Yes, I'll guard any barrier with a verbose check. The barriers will only be used if the timings are being reported.

Reply:
OK.  Is there other verbose output that would make the timers bad?  That is, when verbose = 1, the timers will be reported and the barriers will make the timers good.  But will there also be copious output that makes the timers bad?